### PR TITLE
Functional Refactoring

### PR DESCRIPTION
**Pull request by cloudRoutine**
- added maybe builder
- cleaned up a few functions
- refactored record methods into module functions for pipelining

### CODE REVIEW

File: build.fsx [hunk 1]
```diff
@@ -165,6 +165,21 @@ Target "RunTests" (fun _ ->
             OutputFile = "TestResults.xml" })
 )
 
+Target "QuickTest" (fun _ ->
+
+    !! "C:\Users\Jared\Github\Forks\Paket\src\Paket.Core\Paket.Core.fsproj"
```
Comment (Krzysztof-Cieslak) [inline build.fsx]: Hard-coded path? ;)

Comment (cloudRoutine) [inline build.fsx]: hahaha whoops, nice catch

**Comment by forki** [comment]
why did you add the maybe builder? it's not used

**Comment by cloudRoutine** [comment]
```fsharp

        let parseCondition (data:System.Text.StringBuilder) (input:string) index =
            maybe{
                if input.Length <= index then return! None else 
                data.Clear () |> ignore
                let! left, index = parseWord data input index false
                data.Clear () |> ignore
                let! comp, index = parseComparison data input index
                data.Clear () |> ignore
                let! right, index = parseWord data input index false 
                return (left, comp, right, index)
            }
```
https://github.com/cloudRoutine/Paket/blob/polish/src/Paket.Core/ProjectFile.fs#L298-L308

**Comment by forki** [comment]
it's not in this diff right? And is it really woth it?

**Comment by cloudRoutine** [comment]
It is in it, for some reason it's getting collapsed/hidden in the web UI. Honestly I thought I'd be using it more than I did.

**Comment by forki** [comment]
so the question is if we really want to bring that in. TBH I don't think t that big benefit.

**Comment by cloudRoutine** [comment]
this is what it was before

**Comment by forki** [comment]
how about Option.map? that's already avalaible

**Comment by cloudRoutine** [comment]
Is there anything else I need to address?

File: build.fsx [hunk 2]
```diff
@@ -165,6 +165,21 @@ Target "RunTests" (fun _ ->
...
+
+    !! "src\Paket.Core\Paket.Core.fsproj"    
+    |> MSBuildRelease "" "Rebuild"
+    |> ignore
+
+    !! testAssemblies
+    |> NUnit (fun p ->
+        { p with
+            DisableShadowCopy = true
+            ToolPath = "packages/test/NUnit.Runners.Net4/tools"
```
Comment (forki) [inline build.fsx]: I think you can remove that line. the tool should be discovered

File: tests/Paket.Bootstrapper.Tests/Paket.Bootstrapper.Tests.csproj [hunk 3]
```diff
@@ -58,4 +58,7 @@
       <Paket>True</Paket>
     </Reference>
   </ItemGroup>
-</Project>
+  <ItemGroup>
+    <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
```
Comment (forki) [inline tests/Paket.Bootstrapper.Tests/Paket.Bootstrapper.Tests.csproj]: what's this?

Comment (cloudRoutine) [inline tests/Paket.Bootstrapper.Tests/Paket.Bootstrapper.Tests.csproj]: it gets added every time I run the `RunTests` target

**Comment by forki** [comment]
thanks for helping

**Comment by forki** [comment]
oups. what's that tests/Playlists/paket.tests.playlist thing?

**Comment by cloudRoutine** [comment]
It's for the visual studio test runner so people can run just the tests for paket.core easily

**Comment by forki** [comment]
btw: we're still using F# 3 compiler. 

    normalizePath lib |> FileInfo

is not working on my machine.

**Comment by forki** [comment]
but don't worry. easy to fix for me

**Pull request merged**